server/src/main/java/org/opensearch/snapshots/SnapshotsService.java

```diff
@@ -2230,16 +2230,52 @@ private void deleteSnapshotsFromRepository(
             assert currentlyFinalizing.contains(deleteEntry.repository());
             final List<SnapshotId> snapshotIds = deleteEntry.getSnapshots();
             assert deleteEntry.state() == SnapshotDeletionsInProgress.State.STARTED : "incorrect state for entry [" + deleteEntry + "]";
-            repositoriesService.repository(deleteEntry.repository())
-                .deleteSnapshots(
-                    snapshotIds,
-                    repositoryData.getGenId(),
-                    minCompatibleVersion(minNodeVersion, repositoryData, snapshotIds),
-                    ActionListener.wrap(updatedRepoData -> {
-                        logger.info("snapshots {} deleted", snapshotIds);
-                        removeSnapshotDeletionFromClusterState(deleteEntry, null, updatedRepoData);
-                    }, ex -> removeSnapshotDeletionFromClusterState(deleteEntry, ex, repositoryData))
-                );
+            final Repository repository = repositoriesService.repository(deleteEntry.repository());
+            StepListener<Boolean> checkForShallowSnapshotStep = new StepListener<Boolean>();
+            threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(new AbstractRunnable() {
+                @Override
+                public void onFailure(Exception e) {
+                    checkForShallowSnapshotStep.onFailure(e);
+                }
+
+                @Override
+                protected void doRun() {
+                    boolean cleanupRemoteStoreLockFiles = false;
+                    final List<SnapshotId> allSnapshotIds = repositoryData.getSnapshotIds().stream().collect(Collectors.toList());
+                    for (SnapshotId snapshotId : allSnapshotIds) {
+                        if (Boolean.TRUE.equals(repository.getSnapshotInfo(snapshotId).isRemoteStoreIndexShallowCopyEnabled())) {
+                            cleanupRemoteStoreLockFiles = true;
+                            break;
+                        }
+                    }
+                    checkForShallowSnapshotStep.onResponse(cleanupRemoteStoreLockFiles);
+                }
+            });
+
+            checkForShallowSnapshotStep.whenComplete(cleanupRemoteStoreLockFiles -> {
+                if (cleanupRemoteStoreLockFiles) {
+                    repository.deleteSnapshotsAndReleaseLockFiles(
```

Comment:
>Then in such cases it will end up removing the blobs but not releasing the locks. But I don't expect to enable/disable this flag frequently 

should we think about disabling toggling this flag ?

Reply:
Currently RepoMetadata doesn't support any static sort of setting so we kept it dynamic. However, disabling the toggle of this flag will not help with all the cases as well where some stale blobs may be left in snapshot layer. To deal with this, snapshot layer has a separate `cleanup` API. We will be improving that API to take care of stale blobs for shallow snapshot as well.  

Also with the [recent change to ignore the shallow snapshot blobs from deletion](https://github.com/opensearch-project/OpenSearch/pull/7497/files#diff-0c75016719f0865438856949c6ccf7b3aa0cd121456bbe3826710a422537a736R3168) in such edge cases will prevent lock file leaks. So later a monitor to detect lock file leak and the cleanup API invocation for any such stale blobs will be useful to deal with it.

Reply:
>  some stale blobs may be left in snapshot layer.

Aren't we failing if lock or dat file deletion fails ? As per my understanding we will not have stale blobs in any case .  

> Currently RepoMetadata doesn't support any static sort of setting so we kept it dynamic. However, disabling the toggle of this flag will not help with all the cases as well where some stale blobs may be left in snapshot layer.  

Agreed . For now , it might be okay to assume that is static and later make changes to make it so before 2.10 .

Reply:
> Aren't we failing if lock or dat file deletion fails ? As per my understanding we will not have stale blobs in any case

Yes it fails the delete API if the lock file release or deletion of blobs fails. But before these operation the RepositoryData is already updated to reflect that snapshot is deleted. So any failure later will leave the stale blobs which will be taken care by either future `delete` snapshot call or `cleanup` api call. Similarly, it will be handled for shallow snapshot case too. Disabling the toggle of the flag will not help to solve the problem of stale blobs that will still remain. So the approach is let's keep the flag dynamic (i.e. as is), have a monitor to detect lock file leaks (this will help to catch any other issues not only with delete flow) and then use the cleanup API to take care of deleting stale blobs and lock files.